sdk/core/Azure.Core/src/Shared/AsyncLockWithValue.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Primitive that combines async lock and value cache
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal sealed class AsyncLockWithValue<T>
```

Comment:
This type is a copy of type from [Azure.Identity](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/AsyncLockWithValue.cs) with added `HasValue` property and `TryGetValue` method.

Reply:
Assuming we make this change, we should change Identity's reference to this shared source version also.

Reply:
We can do that.